query-algebrizer/src/types.rs

```diff
@@ -194,6 +194,88 @@ pub enum ColumnConstraint {
     HasType(TableAlias, ValueType),
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraintOrAlternation {
+    Constraint(ColumnConstraint),
+    Alternation(ColumnAlternation),
+}
+
+impl From<ColumnConstraint> for ColumnConstraintOrAlternation {
+    fn from(thing: ColumnConstraint) -> Self {
+        ColumnConstraintOrAlternation::Constraint(thing)
+    }
+}
+
+/// An ordered collection of constraints that must all match.
```

Comment:
Consider: "The `ColumnIntersection` constraint is satisfied if all of the given constraints are satisfied."  What happens if the underlying collection is empty?  Usually `and([])` is `true`, is that what you want?